### PR TITLE
add custom metrics credential CURD related functions for golang apiserver

### PR DESCRIPTION
**Pull request by qibobo**

### CODE REVIEW

File: src/autoscaler/api/brokerserver/broker_handler.go [hunk 1]
```diff
@@ -136,32 +148,80 @@ func (h *BrokerHandler) BindServiceInstance(w http.ResponseWriter, r *http.Reque
...
 		return
 	}
 
+	err = h.bindingdb.CreateServiceBinding(bindingId, instanceId, body.AppID)
+	if err != nil {
+		if err == db.ErrAlreadyExists {
+			h.logger.Error("failed to create binding: binding already exists", err, lager.Data{"appId": body.AppID})
+			handlers.WriteJSONResponse(w, http.StatusConflict, models.ErrorResponse{
+				Code:    "Conflict",
+				Message: "An autoscaler service instance is already bound to application ${applicationId}. Multiple bindings are not supported."})
```
Comment (cdlliuy) [inline src/autoscaler/api/brokerserver/broker_handler.go]: Where is  the variable {applicationId} defined?

File: src/autoscaler/api/brokerserver/broker_handler.go [hunk 2]
```diff
@@ -136,32 +148,80 @@ func (h *BrokerHandler) BindServiceInstance(w http.ResponseWriter, r *http.Reque
...
+			Message: "Error creating service binding"})
+		return
+	}
+	cred, err := custom_metrics_cred_helper.CreateCustomMetricsCredential(body.AppID, h.policydb, custom_metrics_cred_helper.MaxRetry)
+	if err != nil {
+		//revert binding creating
+		h.logger.Error("failed to create custom metrics credential", err, lager.Data{"appId": body.AppID})
+		err = h.bindingdb.DeleteServiceBindingByAppId(body.AppID)
+		if err != nil {
+			h.logger.Error("revert binding due to failed to create custom metrics credential", err, lager.Data{"appId": body.AppID})
```
Comment (cdlliuy) [inline src/autoscaler/api/brokerserver/broker_handler.go]: The log msg is not proper here.

File: src/autoscaler/api/brokerserver/broker_handler.go [hunk 3]
```diff
@@ -179,33 +249,20 @@ func (h *BrokerHandler) BindServiceInstance(w http.ResponseWriter, r *http.Reque
...
-	}
-	w.WriteHeader(http.StatusCreated)
-	w.Write(nil)
+	handlers.WriteJSONResponse(w, http.StatusCreated, models.CredentialResponse{
+		Credentials: models.Credentials{
+			CustomMetrics: models.CustomMetrics{
+				CustomMetricCredentials: cred,
+				URL:                     h.conf.MetricsForwarder.MetricsForwarderUrl,
+			},
+		},
```
Comment (cdlliuy) [inline src/autoscaler/api/brokerserver/broker_handler.go]: @boyang9527   @silvestre , could you share your insight here ?    The current binding credential format  looks complex ...
```
"credentials":{
  "custom_metrics": {
     "username": xxx,
     "password": xxx,
     "url:xxx
  }
}
```

File: src/autoscaler/api/brokerserver/broker_handler.go [hunk 4]
```diff
@@ -41,6 +43,7 @@ func NewBrokerHandler(logger lager.Logger, conf *config.Config, bindingdb db.Bin
 func (h *BrokerHandler) GetBrokerCatalog(w http.ResponseWriter, r *http.Request, vars map[string]string) {
 	catalog, err := ioutil.ReadFile(h.conf.CatalogPath)
 	if err != nil {
+		h.logger.Error("failed to read request body when get catalog", err)
```
Comment (cdlliuy) [inline src/autoscaler/api/brokerserver/broker_handler.go]: The log msg here is improper.  Should be "failed to read catalog file"

File: src/autoscaler/api/brokerserver/broker_handler.go (hunk 3 above)
Comment (silvestre) [reply src/autoscaler/api/brokerserver/broker_handler.go]: If possible, this should not change, as we already have that API GA for our customers. If there's a change here we would need maintain the old structure in our fork.

Comment (cdlliuy) [reply src/autoscaler/api/brokerserver/broker_handler.go]: ok， then let us keep it as is

**Pull request closed without merging**